### DISCUSSION TITLE
Changes can't be pushed, PR submitted instead

### DISCUSSION THREAD

**Discussion by keunes** [WeblateOrg/weblate#6895] [Q&A] [answered]
Hello,

I would like to (manually or automatically) push changes once in a while from Weblate to our `translation-files` branch on our repository (https://github.com/keunes/AP-website-copy for testing purposes). Then from that `translation-files` branch in our repo I want to do a PR to our `main` branch once in a while.

However, I just tried to manually push changes from Weblate to the desired branch. But I got an error:
```
To github.com:weblate/AP-website-copy.git
 ! [remote rejected] translation-files -> weblate-antennapod-website-website-translation-branch-pot-file-contribute (shallow update not allowed)
error: failed to push some refs to 'github.com:weblate/AP-website-copy.git'
 (1)
```

I'm not too familiar with git, so I don't really know what this means.

Also, I noticed that Weblate created a Pull Request to the target branch, instead of just pushing:
https://github.com/keunes/AP-website-copy/pull/10

Are the two (error and fact of receiving a PR) related?

(Or can Weblate never push commits directly, without creating a PR? I read [the docs](https://docs.weblate.org/en/latest/admin/continuous.html#push-changes), but I'm still not sure.)

**Comment by nijel** [comment, marked as answer]
If you choose VCS integration GitHub, it creates pull requests. In case you want to push directly, choose Git.

**Reply by keunes** [reply]
Ah, sorry, I see that now in the docs. Thanks for that hint.

I just switched from GitHub to git and clicked 'Push' in the maintenance repository. The components locked because of the following error:
```
fatal: could not read Username for 'https://github.com': terminal prompts disabled
 (128)
```
The Weblate GitHub user is collaborator for the repo, and the Weblate integration (bot/app) has been installed for the repo. I don't see any other settings in the docs that I might need. Any advice?

Side note: When reading the docs, this section suggests that one needs to configure the system if you _do_ want to have pull requests instead of direct pushes: https://docs.weblate.org/en/latest/vcs.html#pushing-changes-to-github-as-pull-requests Feels contradictory with what is written just above (that GH always sends Pull Requests, use git for direct pushing).

**Reply by keunes** [reply]
The Weblate helpdesk very kindly pointed me in the right direction:
> To use direct push, you need to use the SSH link:
> https://docs.weblate.org/en/latest/vcs.html#ssh-repositories

**Reply by nijel** [reply]
I've added such suggestion to the alert now as well: 7f779e156a5c5d0f82b1adeb4384ed6f75f0455e

**Comment by Quardah** [comment]
Hi

I have the reverse issue, I am trying to get a PR from the forked repo to the main repo on pushes, but in my VCS i can only select 'Git' or 'Git with force push'. How can i add the 'Github' option?

Is it a plugin?

I have installed self hosted 4.9.1.

EDIT: I just figured out that you need to input both a username and a PAT in settings.py on the backend to have the VCS github option enabled.

I have not stumbled upon any notice for this in the documentation, and i believe adding this precision could be usefull.

Thank you for your time and happy new year!!

**Reply by nijel** [reply]
I've tried to improve this in 14af78c31f, additional feedback is welcome.

**Reply by Quardah** [reply]
Hi Nijel. Thank you for the quick response and happy new year.

On line 278 where you can read : "You need to configure API credentials to make this work." I would add the following :

You need to configure API credentials in the weblate settings.py file to make this work. (Username and token OR Credentials). Once properly setup you will see the 'GitHub' option appear in the VCS selector in weblate.

And right under you have the correct "seealso".

This would help future users setup their github integration.

Many thanks for your time and efforts you are doing an incredible job my friend.
JPBD

**Reply by nijel** [reply]
Thanks for suggestion, added this in c6b2e7c3d3ed76c9bd9f5c7280585d9da859f522.

**Reply by Quardah** [reply]
Great.

Just a precision i would add 'Either (:setting:`GITHUB_CREDENTIALS`) or (:setting:`GITHUB_USERNAME`) and (:setting:`GITHUB_TOKEN`)' Just because it can work with either, and there are placeholders in the settings.py file for the username and token variable.

**Reply by nijel** [reply]
Using `GITHUB_CREDENTIALS` is preferred as that avoids credentials leak to another server.

**Comment by chimosky** [comment]
@nijel I just faced the same issue and pushes fail on ` Could not push Core-Sugar/sugar: fatal: could not read Username for 'https://github.com': terminal prompts disabled (128) `, the project uses the ssh link to the repo as there's a GH action that automatically opens a PR when there's a push to a particular branch - which is the one I set as the push branch -.

**Reply by nijel** [reply]
Most likely there is a `https` URL configured for pushing in the component configuration.

**Reply by chimosky** [reply]
Fixed that, and it works. Thank you!
